app/src/test/java/org/simple/clinic/settings/changelanguage/ChangeLanguageUpdateTest.kt

```diff
@@ -0,0 +1,96 @@
+import com.spotify.mobius.test.NextMatchers.hasModel
+import com.spotify.mobius.test.NextMatchers.hasNoEffects
+import com.spotify.mobius.test.NextMatchers.hasNoModel
+import com.spotify.mobius.test.UpdateSpec
+import com.spotify.mobius.test.UpdateSpec.assertThatNext
+import junitparams.JUnitParamsRunner
+import junitparams.Parameters
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.simple.clinic.settings.Language
+
+@RunWith(JUnitParamsRunner::class)
+class ChangeLanguageUpdateTest {
+
+  private val defaultModel = ChangeLanguageModel.FETCHING_LANGUAGES
+
+  private val englishIndia = Language(displayName = "English", languageCode = "en_IN")
+  private val hindiIndia = Language(displayName = "हिंदी", languageCode = "hi_IN")
+
+  private val spec = UpdateSpec<ChangeLanguageModel, ChangeLanguageEvent, ChangeLanguageEffect>(ChangeLanguageLogic::update)
+
+  @Test
+  @Parameters(method = "params for initial models when loading selected language")
+  fun `when the selected language is loaded, the model must be updated`(model: ChangeLanguageModel) {
+    spec
+        .given(model)
+        .whenEvent(CurrentSelectedLanguageLoadedEvent(englishIndia))
+        .then(assertThatNext(
+            hasModel(model.withCurrentLanguage(englishIndia)),
+            hasNoEffects()
+        ))
+  }
+
+  @Suppress("Unused")
+  private fun `params for initial models when loading selected language`(): List<ChangeLanguageModel> {
+    return listOf(
+        defaultModel,
+        defaultModel.withSupportedLanguages(listOf(englishIndia, hindiIndia))
+    )
+  }
+
+  @Test
+  @Parameters(method = "params for initial models when loading supported languages")
+  fun `when the supported languages are loaded, the model must be updated`(model: ChangeLanguageModel) {
+    val supportedLanguages = listOf(englishIndia, hindiIndia)
+
+    spec
+        .given(model)
+        .whenEvent(SupportedLanguagesLoadedEvent(supportedLanguages))
+        .then(assertThatNext(
+            hasModel(model.withSupportedLanguages(supportedLanguages)),
+            hasNoEffects()
+        ))
+  }
+
+  @Suppress("Unused")
+  private fun `params for initial models when loading supported languages`(): List<ChangeLanguageModel> {
+    return listOf(
+        defaultModel,
+        defaultModel.withCurrentLanguage(englishIndia)
+    )
+  }
+
+  @Test
+  fun `when the change language event is triggered, the set selected language effect must be emitted`() {
+    val model = defaultModel
+        .withCurrentLanguage(englishIndia)
+        .withSupportedLanguages(listOf(englishIndia, hindiIndia))
+
+    spec
+        .given(model)
+        .whenEvent(ChangeSelectedLanguageEvent(hindiIndia))
+        .then(assertThatNext(
+            hasNoModel(),
+            hasEffects(SetSelectedLanguageEffect(hindiIndia) as ChangeLanguageEffect)
+        ))
+  }
+
+  @Test
+  fun `when the language changed event is triggered, the model must be updated and the activity must be restarted`() {
```

Comment:
The tests are very technical. They should document the feature itself. For instance,

`when the user selects a language, then apply changes and restart the activity` is a better way. The events and effects are the how and implementation details. Those are presented in the test body itself.

When we are maintaining the code base, changes are described in terms of user stories, not implementation details. Can we reword tests to reflect that instead?

Reply:
Makes sense; I'll try and reword them.

Reply:
Done. Please take a look?